crates/vm/levm/src/precompiles.rs

```diff
@@ -95,12 +98,76 @@ pub fn execute_precompile(current_call_frame: &mut CallFrame) -> Result<Bytes, V
     Ok(result)
 }
 
-fn ecrecover(
-    _calldata: &Bytes,
-    _gas_for_call: U256,
-    _consumed_gas: &mut U256,
+fn fill_with_zeros(slice: &[u8]) -> Result<[u8; 128], VMError> {
+    let mut result = [0; 128];
+
+    let n = slice.len().min(128);
+
+    let trimmed_slice: &[u8] = slice.get(..n).unwrap_or_default();
+
+    for i in 0..n {
+        let byte: &mut u8 = result.get_mut(i).ok_or(InternalError::SlicingError)?;
+        *byte = *trimmed_slice.get(i).ok_or(InternalError::SlicingError)?;
+    }
+
+    Ok(result)
+}
+
+pub fn ecrecover(
+    calldata: &Bytes,
+    gas_for_call: U256,
+    consumed_gas: &mut U256,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    // Consume gas
+    *consumed_gas = consumed_gas
+        .checked_add(ECRECOVER_COST.into())
+        .ok_or(PrecompileError::GasConsumedOverflow)?;
+
+    if gas_for_call < *consumed_gas {
+        return Err(VMError::PrecompileError(PrecompileError::NotEnoughGas));
+    }
+
+    // If calldata does not reach the required length, we should fill the rest with zeros
+    let calldata = fill_with_zeros(calldata)?;
+
+    let hash = calldata.get(0..32).ok_or(InternalError::SlicingError)?;
+    let message = Message::parse_slice(hash).map_err(|_| PrecompileError::ParsingInputError)?;
+
+    let v: U256 = calldata
+        .get(32..64)
+        .ok_or(InternalError::SlicingError)?
+        .into();
+
+    if !(v == U256::from(27) || v == U256::from(28)) {
+        return Ok(Bytes::new());
+    }
+
+    let v = u8::try_from(v).map_err(|_| PrecompileError::ParsingInputError)?;
+    let recovery_id = match RecoveryId::parse_rpc(v) {
+        Ok(id) => id,
+        Err(_) => {
+            return Ok(Bytes::new());
+        }
+    };
+
+    let sig = calldata.get(64..128).ok_or(InternalError::SlicingError)?;
+    let signature =
+        Signature::parse_standard_slice(sig).map_err(|_| PrecompileError::ParsingInputError)?;
+
+    let mut public_key = match libsecp256k1::recover(&message, &signature, &recovery_id) {
+        Ok(id) => id,
+        Err(_) => {
+            return Ok(Bytes::new());
+        }
+    }
+    .serialize();
+
+    keccak256(&mut public_key[1..65]);
+
+    let mut output = vec![0u8; 12];
+    output.extend_from_slice(&public_key[13..33]);
```

Comment:
1. I would add a little comment clarifying that to recover an address we do the keccak-256 of the public key and we take the last 20 bytes of that.
2. You do [1..65] because you want to skip the prefix, right? Can you add a little comment about that too?
3. We may have to change the index slicing for `.get()` and return an internal error if necessary. Do you know why clippy didn't detect that?

Reply:
I think that it's because it's slicing a fixed size slice with a size bigger than that. Anyway, done the 3 points in [Add comments and avoid slicing](https://github.com/lambdaclass/ethrex/pull/1511/commits/2cdb001e3bc5e46fdaf51233f1efbc1ce4cb21c7)

Reply:
Ohh yes, that's the reason, I didn't realize. If you want to keep it like it was then it's ok!